File: modules/ApplyMETUncertainties.py
'''
Script to propagate uncertainties on the JEC's and unclustered energy to MET
'''

import ROOT
from TIMBER.Analyzer import *
from TIMBER.Tools.Common import *

def ApplyMETCorrections(self,variation,applyXYshift=True):

    CompileCpp('modules/ApplyMETShiftXY.cc')
    CompileCpp('modules/ApplyMETUncertainties.cc')
    
    # Apply the XY-shift to the nominal MET
    if applyXYshift:
        self.a.Define('Met_MetPhi_XYShifted',f'METXYCorr_Met_MetPhi("{self.setname}","{self.year}",!{str(self.a.isData).lower()},MET_pt,MET_phi,PV_npvs)')
        self.a.Define('MET_pt_XYshifted','Met_MetPhi_XYShifted.first')
        self.a.Define('MET_phi_XYshifted','Met_MetPhi_XYShifted.second')
    else:
        self.a.Define('MET_pt_XYshifted','MET_pt')
        self.a.Define('MET_phi_XYshifted','MET_phi')

    # Now propagate the JEC's to MET, based on the specified variation
    # Variations are : UE_up/down (uncertainty in the unclustered energy propagated to MET), JES_up/down, JER_up/down
    # Jet mass scale/resolution corrections do not affect the pT of jets, so do not need to be accounted for in the MET
    if 'JES' not in variation and 'JER' not in variation:
        # Propagate nominal JES and JER (for MC) to MET
        if self.a.isData:
            self.a.Define('Jet_JER_nom','Jet_pt')

        self.a.Define('MET_corrected',f'propagate_jec_met(Jet_pt,Jet_JES_nom,Jet_JER_nom,Jet_phi,Jet_rawFactor,Jet_muonSubtrFactor,Jet_neEmEF,Jet_chEmEF,MET_pt_XYshifted,MET_phi_XYshifted,{str(self.a.isData).lower()})')

        self.a.Define('MET_px','MET_corrected[0] * cos(MET_corrected[1])')
        self.a.Define('MET_py','MET_corrected[0] * sin(MET_corrected[1])')
        
        if variation == 'UE_up':
            self.a.Define('MET_px_corr','MET_px + MET_MetUnclustEnUpDeltaX')
            self.a.Define('MET_py_corr','MET_py + MET_MetUnclustEnUpDeltaY')

        elif variation == 'UE_down':
            self.a.Define('MET_px_corr','MET_px - MET_MetUnclustEnUpDeltaX')
            self.a.Define('MET_py_corr','MET_py - MET_MetUnclustEnUpDeltaY')

        else:
            self.a.Define('MET_px_corr','MET_px')
            self.a.Define('MET_py_corr','MET_py')

        self.a.Define('MET_pt_corr','sqrt(MET_px_corr*MET_px_corr + MET_py_corr*MET_py_corr)')
        self.a.Define('MET_phi_corr','atan2(MET_py_corr, MET_px_corr)')

    else: # One of the JES/JER variations
        if 'JES' in variation:
            jes = variation
            jer = 'JER_nom'
        elif 'JER' in variation:
            jes = 'JES_nom'
            jer = variation
        # This code is slightly wrong, as I have not yet implemented L1 JEC's to AK4 jets
        self.a.Define('MET_corrected',f'propagate_jec_met(Jet_pt,Jet_{jes},Jet_{jer},Jet_phi,Jet_rawFactor,Jet_muonSubtrFactor,Jet_neEmEF,Jet_chEmEF,MET_pt_XYshifted,MET_phi_XYshifted,{str(self.a.isData).lower()})')

        self.a.Define('MET_pt_corr','MET_corrected[0]')
        self.a.Define('MET_phi_corr','MET_corrected[1]')

File: modules/ApplyMETUncertainties.cc
// Function to propagate JES/JER corrections to MET
// See: https://github.com/cms-nanoAOD/nanoAOD-tools/blob/master/python/postprocessing/modules/jme/jetmetUncertainties.py#L694


RVec<float> propagate_jec_met(RVec<float> jet_pt, RVec<float> jet_jes, RVec<float> jet_jer, RVec<float> jet_phi, RVec<float> jet_rawFactor, 
		              RVec<float> jet_muSubtFactor, RVec<float> neEmEF, RVec<float> chEmEF, float MET_pt, float MET_phi, bool isData) {

    float MET_px = MET_pt * cos(MET_phi);
    float MET_py = MET_pt * sin(MET_phi);

    for (int i=0; i<jet_pt.size(); i++) {
	// Calculated corrected pt (muon-subtracted)
        float jet_pt_noMu = jet_pt[i] * (1 - jet_rawFactor[i]) * (1 - jet_muSubtFactor[i]);
	float muon_pt = jet_pt[i] * (1 - jet_rawFactor[i]) * jet_muSubtFactor[i]; // pt of muons subtracted from the jet

        float jet_pt_noMuL1L2L3 = jet_pt_noMu * jet_jes[i];
        float jet_pt_L1L2L3 = jet_pt_noMuL1L2L3 + muon_pt;

	float jet_pt_fullCorr; // Fully corrected jet pt
        if (isData) {
	    jet_pt_fullCorr = jet_pt_L1L2L3; // No JER smearing applied to MET in data
	}
	else {
	    jet_pt_fullCorr = jet_pt_L1L2L3 * jet_jer[i];
        }

        if (jet_pt_noMuL1L2L3 > 15 && (neEmEF[i] + chEmEF[i]) < 0.9) { // Only apply to MET if corrected pt (w/o muon) is above threshold
            float jet_px_fullCorr = jet_pt_fullCorr * cos(jet_phi[i]);
            float jet_py_fullCorr = jet_pt_fullCorr * sin(jet_phi[i]);

            float jet_px_L1 = jet_pt[i] * cos(jet_phi[i]);
            float jet_py_L1 = jet_pt[i] * sin(jet_phi[i]);

            MET_px = MET_px - (jet_px_fullCorr - jet_px_L1);
            MET_py = MET_py - (jet_py_fullCorr - jet_py_L1);
	}
    }

    float MET_pt_corr = sqrt(MET_px*MET_px + MET_py*MET_py);
    float MET_phi_corr = atan2(MET_py,MET_px);

    return {MET_pt_corr,MET_phi_corr};
}
